{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<style>\n",
    ".tips {\n",
    "       padding-top: 5px;\n",
    "       padding-bottom: 5px;\n",
    "       padding-left: 5px;\n",
    "       padding-right: 5px;\n",
    "       border: 1px dashed #2f6fab;\n",
    "       #color: Black;\n",
    "       background-color: #EEFFEE;\n",
    "       #color: #330066;\n",
    "       #margin-right: 20%\n",
    "}\n",
    "\n",
    ".code {\n",
    "    font-family: Courier;\n",
    "    padding-top: 0px;\n",
    "    padding-bottom: 0px;\n",
    "    padding-left: 5px;\n",
    "    padding-right: 5px;\n",
    "    border: 1px dashed #2f6fab;\n",
    "    background-color: #EEEEEE;\n",
    "    color: #0000BB;\n",
    "    #margin-right: 20%\n",
    "}\n",
    "\n",
    ".solution {\n",
    "            margin-top: 5px;\n",
    "            margin-bottom: 5px;\n",
    "            padding-top: 5px;\n",
    "            padding-bottom: 5px;\n",
    "            padding-left: 5px;\n",
    "            padding-right: 5px;\n",
    "            border: 1px dashed #FFFFFF;\n",
    "            background-color: #EEEEFF;\n",
    "            color: #0000BB;\n",
    "            font-size: 11px;\n",
    "            #margin-right: 20%\n",
    "}\n",
    "\n",
    "\n",
    ".exo {\n",
    "       margin-top: 5px;\n",
    "       margin-bottom: 5px;\n",
    "       #font-style:italic;\n",
    "       padding-top: 5px;\n",
    "       padding-bottom: 5px;\n",
    "       padding-left: 5px;\n",
    "       padding-right: 5px;\n",
    "       #border: 1px solid #9999ff;\n",
    "       #color: Black;\n",
    "       background-color: #FFEEFF;\n",
    "       color: rgb(51, 51,153);\n",
    "       #margin-right: 30%\n",
    "}\n",
    "\n",
    "\n",
    ".tut {\n",
    "       #font-style:italic;\n",
    "       padding-top: 5px;\n",
    "       padding-bottom: 5px;\n",
    "       padding-left: 20px;\n",
    "       #border: 1px solid #9999ff;\n",
    "       #color: Black;\n",
    "       background-color: #BBDDFF;\n",
    "       color: rgb(51, 51,153);\n",
    "       #margin-right: 30%\n",
    "}\n",
    "\n",
    "\n",
    ".answer {\n",
    "       #font-style:italic;\n",
    "       padding-top: 5px;\n",
    "       padding-bottom: 5px;\n",
    "       padding-left: 20px;\n",
    "       #border: 1px solid #9999ff;\n",
    "       #color: Black;\n",
    "       background-color: #DDFFDD;\n",
    "       color: rgb(51, 51,153);\n",
    "       #margin-right: 30%\n",
    "}\n",
    "\n",
    "\n",
    ".attention {\n",
    "    color: #6B0909;\n",
    "/*    font-size: 13px;*/\n",
    "    padding:5px;\n",
    "    background-color: #FFEEEE;\n",
    "    border: 2px solid #B40404;\n",
    "    margin-right: 5%;\n",
    "    margin-left: 5%;\n",
    "}  \n",
    "\n",
    "/*p.attention\n",
    "{\n",
    "  font-size: 12px;\n",
    "  border: 2px solid #B40404;\n",
    "}\n",
    "*/\n",
    "\n",
    "  /* from RSAT course.css */\n",
    "p.attention\n",
    "{\n",
    "  padding:10px;\n",
    "  background-color: #FFEEEE;\n",
    "  border: 2px solid #B40404;\n",
    "  margin-right: 5%;\n",
    "  margin-left: 5%;\n",
    "}\n",
    "\n",
    "\n",
    ".tbw \n",
    "{\n",
    "  font-style:italic;\n",
    "  padding-top: 5px;\n",
    "  padding-bottom: 5px;\n",
    "  padding-left: 5px;\n",
    "  border: 1px solid #9999ff;\n",
    "  color: Red;\n",
    "  background-color: #FFFFBB;\n",
    "  }\n",
    "\n",
    ".protocol \n",
    "{\n",
    "  font-style:plain;\n",
    "  margin-top: 5px;\n",
    "  margin-bottom: 5px;\n",
    "  padding-top: 5px;\n",
    "  padding-bottom: 5px;\n",
    "  padding-left: 5px;\n",
    "  padding-right: 5px;\n",
    "  #border: 1px solid #9999ff;\n",
    "  color: #000055;\n",
    "  background-color: #D8E8FF;\n",
    "  }\n",
    "\n",
    "/* Concept definitions */\n",
    ".concept \n",
    "{ \n",
    "  color: darkgreen;\n",
    "  font-style: italic;\n",
    "  font-weight: bold;\n",
    "  }\n",
    "\n",
    "/* Figure legend */\n",
    ".legend\n",
    "{ \n",
    "  text-align: center;\n",
    "  font-style: italic;\n",
    "  }\n",
    "\n",
    "/* English terms */\n",
    ".en\n",
    "{ \n",
    "  font-style: italic;\n",
    "  }\n",
    "\n",
    "\n",
    "/* software tool, program */\n",
    ".organism\n",
    "{ \n",
    "  font-style: italic;\n",
    "}\n",
    "\n",
    "/* software tool, program */\n",
    ".program\n",
    "{ \n",
    "  font-style: italic;\n",
    "  font-weight: bold;\n",
    "}\n",
    "\n",
    "\n",
    "/* software tool, program */\n",
    ".command\n",
    "{ \n",
    "  font-weight: bold;\n",
    "  color: brown;\n",
    "  }\n",
    "\n",
    "pre { \n",
    "  background-color: #F5F5F5 !important;\n",
    "  padding: 3px !important;\n",
    "  }\n",
    "  \n",
    "/* avoid the prompt */\n",
    "div.prompt {display:none}\n",
    "  </style>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Objectif\n",
    "\n",
    "On se propose de re-implémenter la commande head (que nous avons précédemment lors des cours sur le language BASH). \n",
    "L'objectif est de créer un programme (sous sa forme la plus simple), capable de:\n",
    "\n",
    "> - Ouvrir un fichier tabulé.\n",
    "> - Parcourir  le fichier ligne à ligne.\n",
    "> - Si le numéro de ligne est inférieur ou égal à 10.\n",
    ">   - Imprimer la ligne.\n",
    "\n",
    "## Données\n",
    "\n",
    "### Téléchargement\n",
    "\n",
    "A titre d'exemple, nous utiliserons un fichier gtf. \n",
    "Ce fichier qui contient les coordonnées des exons de 5000 transcrits dans le génome humain peut être téléchargé ici: [hg38_5k.gtf](../data/hg38_5k.gtf)\n",
    "\n",
    "### Format\n",
    "\n",
    "Le format GTF est un fichier **tabulé** (i.e. dont les colonnes sont séparées par des tabulations) qui contient des informations concernant des éléments génomiques (souvent des transcripts et leurs exons). Les informations portées par les colonnes sont les suivantes:\n",
    "\n",
    "> - **Colonne 1** - Nom du **chromosome**. \n",
    "> - **Colonne 2** - **Source** de l'annotation.\n",
    "> - **Colonne 3** - **Type d'élément** (exon, CDS, 5'UTR, 3'UTR,...).\n",
    "> - **Colonne 4** - **Position de départ** (Start).\n",
    "> - **Colonne 5** - **Position de fin** (End).\n",
    "> - **Colonne 6** - **Score**.\n",
    "> - **Colonne 7** - Le **brin** (+ ou -).\n",
    "> - **Colonne 8** - **Frame** (0,1,2). '0' indique que la première base de l'élément est la première base d'un codon, '1' que la deuxième base est la première base d'un codon, etc...\n",
    "> - **Colonne 9** - **Attribute**. Une liste de couples clefs/valeurs séparées par un point virgule."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "<pre>\n",
    "1       ensembl exon    182393  182746  .       +       .       gene_id \"ENSG00000279928\"; transcript_id \"ENST00000624431\";\n",
    "1       ensembl exon    183114  183240  .       +       .       gene_id \"ENSG00000279928\"; transcript_id \"ENST00000624431\";\n",
    "1       ensembl exon    183922  184158  .       +       .       gene_id \"ENSG00000279928\"; transcript_id \"ENST00000624431\";\n",
    "1       ensembl exon    195263  195411  .       -       .       gene_id \"ENSG00000279457\"; transcript_id \"ENST00000623083\";\n",
    "1       ensembl exon    925741  925800  .       +       .       gene_id \"ENSG00000187634\"; transcript_id \"ENST00000616016\";\n",
    "1       ensembl exon    925741  925800  .       +       .       gene_id \"ENSG00000187634\"; transcript_id \"ENST00000618181\";\n",
    "1       ensembl exon    925741  925800  .       +       .       gene_id \"ENSG00000187634\"; transcript_id \"ENST00000618323\";\n",
    "1       ensembl exon    925741  925800  .       +       .       gene_id \"ENSG00000187634\"; transcript_id \"ENST00000620200\";\n",
    "1       ensembl exon    925922  926013  .       +       .       gene_id \"ENSG00000187634\"; transcript_id \"ENST00000616016\";\n",
    "1       ensembl exon    925922  926013  .       +       .       gene_id \"ENSG00000187634\"; transcript_id \"ENST00000618181\";\n",
    "1       ensembl exon    925922  926013  .       +       .       gene_id \"ENSG00000187634\"; transcript_id \"ENST00000618323\";\n",
    "</pre>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Solutions\n",
    "\n",
    "On présente ci-dessous trois solutions en basant notre approche sur une boucle **for** (2 solutions) ou sur une boucle **while** (une solution). Les deux solutions sont présentées ci-dessous. Le fichier à lire est toujours refGene_hg38.txt (*cf* exercice précédent).\n",
    "\n",
    "### Avec une boucle for"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Solution 1\n",
    "\n",
    "Cette solution fonctionne mais elle est longue puisque l'on parcourra l'ensemble du fichier en faisant un test sur chacune des lignes (on ne s'arrête pas à la dixième ligne). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\thavana\texon\t365389\t365692\t.\t-\t.\ttranscript_version \"1\"; exon_number \"2\"; havana_gene \"OTTHUMG00000156968\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"lincRNA\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000237094\"; exon_id \"ENSE00001656291\"; gene_version \"11\"; transcript_id \"ENST00000431321\"; gene_biotype \"lincRNA\"; havana_transcript \"OTTHUMT00000007989\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"RP4-669L17.10-012\"; havana_gene_version \"6\"; gene_name \"RP4-669L17.10\";\n",
      "1\thavana\texon\t366019\t366120\t.\t-\t.\ttranscript_version \"1\"; exon_number \"1\"; havana_gene \"OTTHUMG00000156968\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"lincRNA\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000237094\"; exon_id \"ENSE00001720138\"; gene_version \"11\"; transcript_id \"ENST00000431321\"; gene_biotype \"lincRNA\"; havana_transcript \"OTTHUMT00000007989\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"RP4-669L17.10-012\"; havana_gene_version \"6\"; gene_name \"RP4-669L17.10\";\n",
      "1\tensembl\texon\t975205\t976269\t.\t-\t.\ttranscript_version \"3\"; exon_number \"3\"; havana_gene \"OTTHUMG00000040757\"; transcript_support_level \"5\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; tag \"basic\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000187642\"; exon_id \"ENSE00002263633\"; ccds_id \"CCDS76083\"; gene_version \"9\"; transcript_id \"ENST00000433179\"; gene_biotype \"protein_coding\"; exon_version \"2\"; transcript_name \"PERM1-201\"; havana_gene_version \"3\"; gene_name \"PERM1\";\n",
      "1\tensembl\texon\t976499\t976624\t.\t-\t.\ttranscript_version \"3\"; exon_number \"2\"; havana_gene \"OTTHUMG00000040757\"; transcript_support_level \"5\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; tag \"basic\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000187642\"; exon_id \"ENSE00001389758\"; ccds_id \"CCDS76083\"; gene_version \"9\"; transcript_id \"ENST00000433179\"; gene_biotype \"protein_coding\"; exon_version \"1\"; transcript_name \"PERM1-201\"; havana_gene_version \"3\"; gene_name \"PERM1\";\n",
      "1\tensembl\texon\t978881\t981029\t.\t-\t.\ttranscript_version \"3\"; exon_number \"1\"; havana_gene \"OTTHUMG00000040757\"; transcript_support_level \"5\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; tag \"basic\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000187642\"; exon_id \"ENSE00003728856\"; ccds_id \"CCDS76083\"; gene_version \"9\"; transcript_id \"ENST00000433179\"; gene_biotype \"protein_coding\"; exon_version \"1\"; transcript_name \"PERM1-201\"; havana_gene_version \"3\"; gene_name \"PERM1\";\n",
      "1\thavana\texon\t1242449\t1242646\t.\t-\t.\ttranscript_version \"1\"; exon_number \"2\"; havana_gene \"OTTHUMG00000001412\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"retained_intron\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000184163\"; exon_id \"ENSE00001853874\"; gene_version \"3\"; transcript_id \"ENST00000478606\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000005166\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"FAM132A-005\"; havana_gene_version \"6\"; gene_name \"FAM132A\";\n",
      "1\thavana\texon\t1242835\t1243158\t.\t-\t.\ttranscript_version \"1\"; exon_number \"1\"; havana_gene \"OTTHUMG00000001412\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"retained_intron\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000184163\"; exon_id \"ENSE00001951859\"; gene_version \"3\"; transcript_id \"ENST00000478606\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000005166\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"FAM132A-005\"; havana_gene_version \"6\"; gene_name \"FAM132A\";\n",
      "1\thavana\texon\t1312787\t1312949\t.\t-\t.\ttranscript_version \"5\"; exon_number \"7\"; havana_gene \"OTTHUMG00000003330\"; transcript_support_level \"5 (assigned to previous version 4)\"; transcript_source \"havana\"; transcript_biotype \"nonsense_mediated_decay\"; tag \"mRNA_start_NF\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000127054\"; exon_id \"ENSE00003654424\"; gene_version \"18\"; transcript_id \"ENST00000531377\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000383033\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"CPSF3L-042\"; havana_gene_version \"12\"; gene_name \"CPSF3L\";\n",
      "1\thavana\texon\t1313035\t1313124\t.\t-\t.\ttranscript_version \"5\"; exon_number \"6\"; havana_gene \"OTTHUMG00000003330\"; transcript_support_level \"5 (assigned to previous version 4)\"; transcript_source \"havana\"; transcript_biotype \"nonsense_mediated_decay\"; tag \"mRNA_start_NF\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000127054\"; exon_id \"ENSE00003616764\"; gene_version \"18\"; transcript_id \"ENST00000531377\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000383033\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"CPSF3L-042\"; havana_gene_version \"12\"; gene_name \"CPSF3L\";\n",
      "1\thavana\texon\t1313509\t1313592\t.\t-\t.\ttranscript_version \"5\"; exon_number \"5\"; havana_gene \"OTTHUMG00000003330\"; transcript_support_level \"5 (assigned to previous version 4)\"; transcript_source \"havana\"; transcript_biotype \"nonsense_mediated_decay\"; tag \"mRNA_start_NF\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000127054\"; exon_id \"ENSE00003594177\"; gene_version \"18\"; transcript_id \"ENST00000531377\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000383033\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"CPSF3L-042\"; havana_gene_version \"12\"; gene_name \"CPSF3L\";\n"
     ]
    }
   ],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "# La ligne précédente permet d'utiliser des caractères accentués\n",
    "# dans les commentaires\n",
    "\n",
    "# Pensez à adapter le chemin vers le fichier.\n",
    "# Ici, il est dans le dossier data présent dans le\n",
    "# répertoire supérieur \n",
    "file_handler = open(\"../data/hg38_5k.gtf\", \"r\")\n",
    "\n",
    "# On définit une variable line_number\n",
    "# qui contiendra le numéro de la ligne courante\n",
    "line_number = 0\n",
    "\n",
    "# On utilise une boucle for pour \n",
    "# lire des lignes renvoyées par l'objet\n",
    "# file_handler\n",
    "for line in file_handler:\n",
    "        \n",
    "        # On incrémente la variable line_number\n",
    "        # On est maintenant à la ligne 'line_number'\n",
    "        line_number = line_number + 1\n",
    "        \n",
    "        # On test si l'expression suivante (line_number <= 10)\n",
    "        # est vrai (i.e si elle est vérifiée).\n",
    "        if line_number <= 10:\n",
    "                # Si la condition est vrai on retire le caractère \n",
    "                # \"\\n\" à droite. En effet, print produira lui même un \n",
    "                # caractère \"\\n\".\n",
    "                line = line.rstrip(\"\\n\")\n",
    "                print(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    Le temps d'execution est ici de **0.184 secondes**. Vous pouvez le tester sur votre machine en utilisant la ligne commande suivante dans le terminal:\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "$ time python  head.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Solution 2\n",
    "\n",
    "Toujours avec une boucle for mais cette fois ci nous utilisons le **module sys** afin d'importer la fonction **exit()** permettant d'arrêter le programme si la condition n'est pas vérifiée."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\thavana\texon\t365389\t365692\t.\t-\t.\ttranscript_version \"1\"; exon_number \"2\"; havana_gene \"OTTHUMG00000156968\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"lincRNA\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000237094\"; exon_id \"ENSE00001656291\"; gene_version \"11\"; transcript_id \"ENST00000431321\"; gene_biotype \"lincRNA\"; havana_transcript \"OTTHUMT00000007989\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"RP4-669L17.10-012\"; havana_gene_version \"6\"; gene_name \"RP4-669L17.10\";\n",
      "1\thavana\texon\t366019\t366120\t.\t-\t.\ttranscript_version \"1\"; exon_number \"1\"; havana_gene \"OTTHUMG00000156968\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"lincRNA\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000237094\"; exon_id \"ENSE00001720138\"; gene_version \"11\"; transcript_id \"ENST00000431321\"; gene_biotype \"lincRNA\"; havana_transcript \"OTTHUMT00000007989\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"RP4-669L17.10-012\"; havana_gene_version \"6\"; gene_name \"RP4-669L17.10\";\n",
      "1\tensembl\texon\t975205\t976269\t.\t-\t.\ttranscript_version \"3\"; exon_number \"3\"; havana_gene \"OTTHUMG00000040757\"; transcript_support_level \"5\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; tag \"basic\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000187642\"; exon_id \"ENSE00002263633\"; ccds_id \"CCDS76083\"; gene_version \"9\"; transcript_id \"ENST00000433179\"; gene_biotype \"protein_coding\"; exon_version \"2\"; transcript_name \"PERM1-201\"; havana_gene_version \"3\"; gene_name \"PERM1\";\n",
      "1\tensembl\texon\t976499\t976624\t.\t-\t.\ttranscript_version \"3\"; exon_number \"2\"; havana_gene \"OTTHUMG00000040757\"; transcript_support_level \"5\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; tag \"basic\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000187642\"; exon_id \"ENSE00001389758\"; ccds_id \"CCDS76083\"; gene_version \"9\"; transcript_id \"ENST00000433179\"; gene_biotype \"protein_coding\"; exon_version \"1\"; transcript_name \"PERM1-201\"; havana_gene_version \"3\"; gene_name \"PERM1\";\n",
      "1\tensembl\texon\t978881\t981029\t.\t-\t.\ttranscript_version \"3\"; exon_number \"1\"; havana_gene \"OTTHUMG00000040757\"; transcript_support_level \"5\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; tag \"basic\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000187642\"; exon_id \"ENSE00003728856\"; ccds_id \"CCDS76083\"; gene_version \"9\"; transcript_id \"ENST00000433179\"; gene_biotype \"protein_coding\"; exon_version \"1\"; transcript_name \"PERM1-201\"; havana_gene_version \"3\"; gene_name \"PERM1\";\n",
      "1\thavana\texon\t1242449\t1242646\t.\t-\t.\ttranscript_version \"1\"; exon_number \"2\"; havana_gene \"OTTHUMG00000001412\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"retained_intron\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000184163\"; exon_id \"ENSE00001853874\"; gene_version \"3\"; transcript_id \"ENST00000478606\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000005166\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"FAM132A-005\"; havana_gene_version \"6\"; gene_name \"FAM132A\";\n",
      "1\thavana\texon\t1242835\t1243158\t.\t-\t.\ttranscript_version \"1\"; exon_number \"1\"; havana_gene \"OTTHUMG00000001412\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"retained_intron\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000184163\"; exon_id \"ENSE00001951859\"; gene_version \"3\"; transcript_id \"ENST00000478606\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000005166\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"FAM132A-005\"; havana_gene_version \"6\"; gene_name \"FAM132A\";\n",
      "1\thavana\texon\t1312787\t1312949\t.\t-\t.\ttranscript_version \"5\"; exon_number \"7\"; havana_gene \"OTTHUMG00000003330\"; transcript_support_level \"5 (assigned to previous version 4)\"; transcript_source \"havana\"; transcript_biotype \"nonsense_mediated_decay\"; tag \"mRNA_start_NF\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000127054\"; exon_id \"ENSE00003654424\"; gene_version \"18\"; transcript_id \"ENST00000531377\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000383033\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"CPSF3L-042\"; havana_gene_version \"12\"; gene_name \"CPSF3L\";\n",
      "1\thavana\texon\t1313035\t1313124\t.\t-\t.\ttranscript_version \"5\"; exon_number \"6\"; havana_gene \"OTTHUMG00000003330\"; transcript_support_level \"5 (assigned to previous version 4)\"; transcript_source \"havana\"; transcript_biotype \"nonsense_mediated_decay\"; tag \"mRNA_start_NF\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000127054\"; exon_id \"ENSE00003616764\"; gene_version \"18\"; transcript_id \"ENST00000531377\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000383033\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"CPSF3L-042\"; havana_gene_version \"12\"; gene_name \"CPSF3L\";\n",
      "1\thavana\texon\t1313509\t1313592\t.\t-\t.\ttranscript_version \"5\"; exon_number \"5\"; havana_gene \"OTTHUMG00000003330\"; transcript_support_level \"5 (assigned to previous version 4)\"; transcript_source \"havana\"; transcript_biotype \"nonsense_mediated_decay\"; tag \"mRNA_start_NF\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000127054\"; exon_id \"ENSE00003594177\"; gene_version \"18\"; transcript_id \"ENST00000531377\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000383033\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"CPSF3L-042\"; havana_gene_version \"12\"; gene_name \"CPSF3L\";\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "To exit: use 'exit', 'quit', or Ctrl-D.\n"
     ]
    }
   ],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "# La ligne précédente permet d'utiliser des caractères accentués\n",
    "# dans les commentaires\n",
    "\n",
    "import sys # on importe le module sys\n",
    "\n",
    "# Pensez à adapter le chemin vers le fichier.\n",
    "# Ici, il est dans le dossier data présent dans le\n",
    "# répertoire supérieur \n",
    "file_handler = open(\"../data/hg38_5k.gtf\", \"r\")\n",
    "\n",
    "# On définit une variable line_number\n",
    "# qui contiendra le numéro de la ligne courante\n",
    "line_number = 0\n",
    "\n",
    "# On utilise une boucle for pour \n",
    "# lire des lignes renvoyées par l'objet\n",
    "# file_handler\n",
    "for line in file_handler:\n",
    "        \n",
    "        # On incrémente la variable line_number\n",
    "        # On est maintenant à la ligne 'line_number'\n",
    "        line_number = line_number + 1\n",
    "        \n",
    "        # On test si l'expression suivante (line_number <= 10)\n",
    "        # est vrai (i.e si elle est vérifiée).\n",
    "        if line_number <= 10:\n",
    "                # Si la condition est vrai on retire le caractère \n",
    "                # \"\\n\" à droite. En effet, print produira lui même un \n",
    "                # caractère \"\\n\".\n",
    "                line = line.rstrip(\"\\n\")\n",
    "                print(line)\n",
    "        else:\n",
    "            # Si on est au delà de la ligne 10\n",
    "            # on quitte le programme (on ne lit\n",
    "            # pas les lignes suivantes)\n",
    "            sys.exit()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Le temps d'exécution est ici de **0.022 secondes soit un gain de 0.162 secondes**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Solution 3\n",
    "\n",
    "Toujours avec une boucle for mais cette fois ci nous utilisons le mot clef break permettant d'arrêter la boucle si la condition n'est pas vérifiée."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\thavana\texon\t365389\t365692\t.\t-\t.\ttranscript_version \"1\"; exon_number \"2\"; havana_gene \"OTTHUMG00000156968\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"lincRNA\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000237094\"; exon_id \"ENSE00001656291\"; gene_version \"11\"; transcript_id \"ENST00000431321\"; gene_biotype \"lincRNA\"; havana_transcript \"OTTHUMT00000007989\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"RP4-669L17.10-012\"; havana_gene_version \"6\"; gene_name \"RP4-669L17.10\";\n",
      "1\thavana\texon\t366019\t366120\t.\t-\t.\ttranscript_version \"1\"; exon_number \"1\"; havana_gene \"OTTHUMG00000156968\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"lincRNA\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000237094\"; exon_id \"ENSE00001720138\"; gene_version \"11\"; transcript_id \"ENST00000431321\"; gene_biotype \"lincRNA\"; havana_transcript \"OTTHUMT00000007989\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"RP4-669L17.10-012\"; havana_gene_version \"6\"; gene_name \"RP4-669L17.10\";\n",
      "1\tensembl\texon\t975205\t976269\t.\t-\t.\ttranscript_version \"3\"; exon_number \"3\"; havana_gene \"OTTHUMG00000040757\"; transcript_support_level \"5\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; tag \"basic\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000187642\"; exon_id \"ENSE00002263633\"; ccds_id \"CCDS76083\"; gene_version \"9\"; transcript_id \"ENST00000433179\"; gene_biotype \"protein_coding\"; exon_version \"2\"; transcript_name \"PERM1-201\"; havana_gene_version \"3\"; gene_name \"PERM1\";\n",
      "1\tensembl\texon\t976499\t976624\t.\t-\t.\ttranscript_version \"3\"; exon_number \"2\"; havana_gene \"OTTHUMG00000040757\"; transcript_support_level \"5\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; tag \"basic\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000187642\"; exon_id \"ENSE00001389758\"; ccds_id \"CCDS76083\"; gene_version \"9\"; transcript_id \"ENST00000433179\"; gene_biotype \"protein_coding\"; exon_version \"1\"; transcript_name \"PERM1-201\"; havana_gene_version \"3\"; gene_name \"PERM1\";\n",
      "1\tensembl\texon\t978881\t981029\t.\t-\t.\ttranscript_version \"3\"; exon_number \"1\"; havana_gene \"OTTHUMG00000040757\"; transcript_support_level \"5\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; tag \"basic\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000187642\"; exon_id \"ENSE00003728856\"; ccds_id \"CCDS76083\"; gene_version \"9\"; transcript_id \"ENST00000433179\"; gene_biotype \"protein_coding\"; exon_version \"1\"; transcript_name \"PERM1-201\"; havana_gene_version \"3\"; gene_name \"PERM1\";\n",
      "1\thavana\texon\t1242449\t1242646\t.\t-\t.\ttranscript_version \"1\"; exon_number \"2\"; havana_gene \"OTTHUMG00000001412\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"retained_intron\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000184163\"; exon_id \"ENSE00001853874\"; gene_version \"3\"; transcript_id \"ENST00000478606\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000005166\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"FAM132A-005\"; havana_gene_version \"6\"; gene_name \"FAM132A\";\n",
      "1\thavana\texon\t1242835\t1243158\t.\t-\t.\ttranscript_version \"1\"; exon_number \"1\"; havana_gene \"OTTHUMG00000001412\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"retained_intron\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000184163\"; exon_id \"ENSE00001951859\"; gene_version \"3\"; transcript_id \"ENST00000478606\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000005166\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"FAM132A-005\"; havana_gene_version \"6\"; gene_name \"FAM132A\";\n",
      "1\thavana\texon\t1312787\t1312949\t.\t-\t.\ttranscript_version \"5\"; exon_number \"7\"; havana_gene \"OTTHUMG00000003330\"; transcript_support_level \"5 (assigned to previous version 4)\"; transcript_source \"havana\"; transcript_biotype \"nonsense_mediated_decay\"; tag \"mRNA_start_NF\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000127054\"; exon_id \"ENSE00003654424\"; gene_version \"18\"; transcript_id \"ENST00000531377\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000383033\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"CPSF3L-042\"; havana_gene_version \"12\"; gene_name \"CPSF3L\";\n",
      "1\thavana\texon\t1313035\t1313124\t.\t-\t.\ttranscript_version \"5\"; exon_number \"6\"; havana_gene \"OTTHUMG00000003330\"; transcript_support_level \"5 (assigned to previous version 4)\"; transcript_source \"havana\"; transcript_biotype \"nonsense_mediated_decay\"; tag \"mRNA_start_NF\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000127054\"; exon_id \"ENSE00003616764\"; gene_version \"18\"; transcript_id \"ENST00000531377\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000383033\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"CPSF3L-042\"; havana_gene_version \"12\"; gene_name \"CPSF3L\";\n",
      "1\thavana\texon\t1313509\t1313592\t.\t-\t.\ttranscript_version \"5\"; exon_number \"5\"; havana_gene \"OTTHUMG00000003330\"; transcript_support_level \"5 (assigned to previous version 4)\"; transcript_source \"havana\"; transcript_biotype \"nonsense_mediated_decay\"; tag \"mRNA_start_NF\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000127054\"; exon_id \"ENSE00003594177\"; gene_version \"18\"; transcript_id \"ENST00000531377\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000383033\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"CPSF3L-042\"; havana_gene_version \"12\"; gene_name \"CPSF3L\";\n"
     ]
    }
   ],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "# La ligne précédente permet d'utiliser des caractères accentués\n",
    "# dans les commentaires\n",
    "\n",
    "# Pensez à adapter le chemin vers le fichier.\n",
    "# Ici, il est dans le dossier data présent dans le\n",
    "# répertoire supérieur \n",
    "file_handler = open(\"../data/hg38_5k.gtf\", \"r\")\n",
    "\n",
    "# On définit une variable line_number\n",
    "# qui contiendra le numéro de la ligne courante\n",
    "line_number = 0\n",
    "\n",
    "# On utilise une boucle for pour \n",
    "# lire des lignes renvoyées par l'objet\n",
    "# file_handler\n",
    "for line in file_handler:\n",
    "        \n",
    "        # On incrémente la variable line_number\n",
    "        # On est maintenant à la ligne 'line_number'\n",
    "        line_number = line_number + 1\n",
    "        \n",
    "        # On test si l'expression suivante (line_number <= 10)\n",
    "        # est vrai (i.e si elle est vérifiée).\n",
    "        if line_number <= 10:\n",
    "                # Si la condition est vrai on retire le caractère \n",
    "                # \"\\n\" à droite. En effet, print produira lui même un \n",
    "                # caractère \"\\n\".\n",
    "                line = line.rstrip(\"\\n\")\n",
    "                print(line)\n",
    "        else:\n",
    "            # Si on est au delà de la ligne 10\n",
    "            # on quitte la boucle\n",
    "            break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Solution 4 (boucle while)\n",
    "\n",
    "On change de point de vue et on utilise une boucle **while**. Le principe est de dire:\n",
    "\n",
    "> - Tant que le numero de ligne est inférieur ou égal à 10\n",
    ">    - Lit une ligne.\n",
    "\n",
    "C'est un peu différent du point de vue de la syntaxe et plus élégant. A noter l'utilisation de la méthode **readline()** disponible pour les objets de type/classe **file**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\thavana\texon\t365389\t365692\t.\t-\t.\ttranscript_version \"1\"; exon_number \"2\"; havana_gene \"OTTHUMG00000156968\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"lincRNA\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000237094\"; exon_id \"ENSE00001656291\"; gene_version \"11\"; transcript_id \"ENST00000431321\"; gene_biotype \"lincRNA\"; havana_transcript \"OTTHUMT00000007989\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"RP4-669L17.10-012\"; havana_gene_version \"6\"; gene_name \"RP4-669L17.10\";\n",
      "1\thavana\texon\t366019\t366120\t.\t-\t.\ttranscript_version \"1\"; exon_number \"1\"; havana_gene \"OTTHUMG00000156968\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"lincRNA\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000237094\"; exon_id \"ENSE00001720138\"; gene_version \"11\"; transcript_id \"ENST00000431321\"; gene_biotype \"lincRNA\"; havana_transcript \"OTTHUMT00000007989\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"RP4-669L17.10-012\"; havana_gene_version \"6\"; gene_name \"RP4-669L17.10\";\n",
      "1\tensembl\texon\t975205\t976269\t.\t-\t.\ttranscript_version \"3\"; exon_number \"3\"; havana_gene \"OTTHUMG00000040757\"; transcript_support_level \"5\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; tag \"basic\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000187642\"; exon_id \"ENSE00002263633\"; ccds_id \"CCDS76083\"; gene_version \"9\"; transcript_id \"ENST00000433179\"; gene_biotype \"protein_coding\"; exon_version \"2\"; transcript_name \"PERM1-201\"; havana_gene_version \"3\"; gene_name \"PERM1\";\n",
      "1\tensembl\texon\t976499\t976624\t.\t-\t.\ttranscript_version \"3\"; exon_number \"2\"; havana_gene \"OTTHUMG00000040757\"; transcript_support_level \"5\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; tag \"basic\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000187642\"; exon_id \"ENSE00001389758\"; ccds_id \"CCDS76083\"; gene_version \"9\"; transcript_id \"ENST00000433179\"; gene_biotype \"protein_coding\"; exon_version \"1\"; transcript_name \"PERM1-201\"; havana_gene_version \"3\"; gene_name \"PERM1\";\n",
      "1\tensembl\texon\t978881\t981029\t.\t-\t.\ttranscript_version \"3\"; exon_number \"1\"; havana_gene \"OTTHUMG00000040757\"; transcript_support_level \"5\"; transcript_source \"ensembl\"; transcript_biotype \"protein_coding\"; tag \"basic\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000187642\"; exon_id \"ENSE00003728856\"; ccds_id \"CCDS76083\"; gene_version \"9\"; transcript_id \"ENST00000433179\"; gene_biotype \"protein_coding\"; exon_version \"1\"; transcript_name \"PERM1-201\"; havana_gene_version \"3\"; gene_name \"PERM1\";\n",
      "1\thavana\texon\t1242449\t1242646\t.\t-\t.\ttranscript_version \"1\"; exon_number \"2\"; havana_gene \"OTTHUMG00000001412\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"retained_intron\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000184163\"; exon_id \"ENSE00001853874\"; gene_version \"3\"; transcript_id \"ENST00000478606\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000005166\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"FAM132A-005\"; havana_gene_version \"6\"; gene_name \"FAM132A\";\n",
      "1\thavana\texon\t1242835\t1243158\t.\t-\t.\ttranscript_version \"1\"; exon_number \"1\"; havana_gene \"OTTHUMG00000001412\"; transcript_support_level \"2\"; transcript_source \"havana\"; transcript_biotype \"retained_intron\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000184163\"; exon_id \"ENSE00001951859\"; gene_version \"3\"; transcript_id \"ENST00000478606\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000005166\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"FAM132A-005\"; havana_gene_version \"6\"; gene_name \"FAM132A\";\n",
      "1\thavana\texon\t1312787\t1312949\t.\t-\t.\ttranscript_version \"5\"; exon_number \"7\"; havana_gene \"OTTHUMG00000003330\"; transcript_support_level \"5 (assigned to previous version 4)\"; transcript_source \"havana\"; transcript_biotype \"nonsense_mediated_decay\"; tag \"mRNA_start_NF\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000127054\"; exon_id \"ENSE00003654424\"; gene_version \"18\"; transcript_id \"ENST00000531377\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000383033\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"CPSF3L-042\"; havana_gene_version \"12\"; gene_name \"CPSF3L\";\n",
      "1\thavana\texon\t1313035\t1313124\t.\t-\t.\ttranscript_version \"5\"; exon_number \"6\"; havana_gene \"OTTHUMG00000003330\"; transcript_support_level \"5 (assigned to previous version 4)\"; transcript_source \"havana\"; transcript_biotype \"nonsense_mediated_decay\"; tag \"mRNA_start_NF\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000127054\"; exon_id \"ENSE00003616764\"; gene_version \"18\"; transcript_id \"ENST00000531377\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000383033\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"CPSF3L-042\"; havana_gene_version \"12\"; gene_name \"CPSF3L\";\n",
      "1\thavana\texon\t1313509\t1313592\t.\t-\t.\ttranscript_version \"5\"; exon_number \"5\"; havana_gene \"OTTHUMG00000003330\"; transcript_support_level \"5 (assigned to previous version 4)\"; transcript_source \"havana\"; transcript_biotype \"nonsense_mediated_decay\"; tag \"mRNA_start_NF\"; gene_source \"ensembl_havana\"; gene_id \"ENSG00000127054\"; exon_id \"ENSE00003594177\"; gene_version \"18\"; transcript_id \"ENST00000531377\"; gene_biotype \"protein_coding\"; havana_transcript \"OTTHUMT00000383033\"; exon_version \"1\"; havana_transcript_version \"1\"; transcript_name \"CPSF3L-042\"; havana_gene_version \"12\"; gene_name \"CPSF3L\";\n"
     ]
    }
   ],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "# La ligne précédente permet d'utiliser des caractères accentués\n",
    "# dans les commentaires\n",
    "\n",
    "# Pensez à adapter le chemin vers le fichier.\n",
    "# Ici, il est dans le dossier data présent dans le\n",
    "# répertoire supérieur \n",
    "file_handler = open(\"../data/hg38_5k.gtf\", \"r\")\n",
    "\n",
    "# On définit une variable line_number\n",
    "# qui contiendra le numéro de la ligne courante\n",
    "line_number = 0\n",
    "\n",
    "while  line_number < 10:\n",
    "\n",
    "        # On lit une ligne du fichier avec \n",
    "        # la méthode readline()\n",
    "        line =  file_handler.readline()\n",
    "        # On retire le caractère \n",
    "        # \"\\n\" à droite. En effet, print produira lui même un \n",
    "        # caractère \"\\n\".        \n",
    "        line = line.rstrip(\"\\n\")\n",
    "        print(line)\n",
    "        \n",
    "        # On ajoute 1 à line_number\n",
    "        # Sinon la condition 'line_number < 10'\n",
    "        # sera toujours vérifiée\n",
    "        line_number = line_number + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Le temps d'exécution est là aussi de **0.022 secondes."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
